File: Java/Leetcode/Heap/TaskScheduler.java
package com.brianprojects.Leetcode.Heap;



public class TaskScheduler {
    public int leastInterval(char[] tasks, int n){
        int[] count = new int[26];
        for(char task:tasks){
            count[task-'A']++;
        }

        int max = 0;
        int cnt = 0;
        for(int i=0;i<count.length;i++){
            max = Math.max(max, count[i]);
        }
        for(int i = 0; i<count.length; i++){
            if(count[i] == max){cnt++;}
        }

        return Math.max(tasks.length, (max-1) *(n+1)+cnt);
    }
}
/*
Queue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
        Queue<int[]> queue = new ArrayDeque<>();
        int[] charCount = new int[26];

        int time = 0;
        for(var ch : tasks){
            charCount[ch-'A'] +=1;
        }
        for(var count: charCount){
            if(count != 0){
                maxHeap.offer(count);
            }
        }

        while(!maxHeap.isEmpty() || !queue.isEmpty()){
            if(maxHeap.isEmpty()){
                time = queue.peek()[0] == 0? time: queue.peek()[1];
            }
            if((!queue.isEmpty() && queue.peek()[1] == time)){
                maxHeap.offer(queue.poll()[0]);
            }
            var next = maxHeap.poll();
            next-=1;
            ++time;
            if(next !=0){
                queue.add(new int[]{next, time+n});
            }

        }
        return time;
 */

File: Python/Leetcode_Python/Heap/TaskScheduler.py
from typing import List
import heapq


def leastInterval(tasks: List[str], n: int) -> int:
    if not tasks:
        return -1

    time = 0
    maxHeap = []
    queue = []
    letters = [0] * 26

    for task in tasks:
        letters[ord(task) - 65] += 1

    for letter in letters:
        if letter != 0:
            heapq.heappush(maxHeap, -letter)

    while True:
        if not maxHeap and not queue:
            break
        if len(queue) != 0 and queue[0][1] <= time:
            temp = queue.pop(0)
            heapq.heappush(maxHeap, temp[0])
        if len(maxHeap) != 0:
            val = heapq.heappop(maxHeap) + 1
            if val != 0:
                queue.append([val, time + n + 1])
        time += 1

    return time
